{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24\n",
      "This is the list for the forward primers['tacccaagaataggcctacc\\n', 'taccaccttgtttgacacca\\n', 'taccaccgatacaaagttct\\n', 'tactcctatttttactccaa\\n', 'tacgaatgaaggaagtttag\\n', 'tacctgttacattctccaag\\n', 'tacgggaataggaaactttc\\n', 'taccgccttaggccgctaaa\\n', 'tacgtttgcagagtctttgt\\n', 'tacttttagggacgtagtag\\n', 'tactgatgttttgcgtatct\\n', 'taccttagaccaaagaggcc\\n']\n",
      "This is the list for the reverse primers['ctatgaagacgaaggaaccc\\n', 'ttacctccaagctgaagcaa\\n', 'tcacttcctgcatctccaag\\n', 'ttacctccaagcacaagaag\\n', 'tcatttccacgacgaaacgg\\n', 'tcacttccatgcacaagatg\\n', 'ctaacatttccaagcagaga\\n', 'ctaagaacgaggcgtccaag\\n', 'ctagtttgggtttggcttcc\\n', 'ctaaaaccgccaagctgatg\\n', 'tcaatcgaacggctgagatt\\n', 'ttaacgccaagcggatgcga\\n']\n",
      "\n",
      "This positions from the forward primers need to be revised because of low temperature: [3, 4, 10]\n",
      "\n",
      "Forward melting points are:[60, 58, 56, 56, 58, 56, 56, 62, 56, 56, 58, 60]\n",
      "Reverse melting points are:[60, 58, 60, 58, 60, 58, 56, 62, 60, 60, 58, 62]\n",
      "\n",
      "Difference between melting temperatures is optimal\n",
      "\n",
      "The primers for each sequence are:\n",
      ">AT1G07530.1 | Symbols: ATGRAS2, GRAS2, SCL14 | SCARECROW-like 14, \"GRAS (GAI, RGA, SCR) 2\", \"ARABIDOPSIS THALIANA GRAS (GAI, RGA, SCR) 2\" | chr1:2313828-2316137 REVERSE LENGTH=2310\n",
      "tacccaagaataggcctacc\n",
      "ctatgaagacgaaggaaccc\n",
      ">AT1G21450.1 | Symbols: SCL1 | SCARECROW-like 1 | chr1:7509721-7511502 FORWARD LENGTH=1782\n",
      "taccaccttgtttgacacca\n",
      "ttacctccaagctgaagcaa\n",
      ">AT1G50420.1 | Symbols: SCL3, SCL-3 | scarecrow-like 3, SCARECROW-LIKE 3 | chr1:18678177-18679625 REVERSE LENGTH=1449\n",
      "taccaccgatacaaagttct\n",
      "tcacttcctgcatctccaag\n",
      ">AT1G50600.2 | Symbols: SCL5 | scarecrow-like 5 | chr1:18737398-18739559 REVERSE LENGTH=1806\n",
      "tactcctgtttttgctccaa\n",
      "ttacctccaagcacaagaag\n",
      ">AT1G55580.1 | Symbols: SCL18, LAS | Lateral Suppressor, SCARECROW-LIKE 18 | chr1:20764106-20765443 FORWARD LENGTH=1338\n",
      "tacgggtgaaggaagtttag\n",
      "tcatttccacgacgaaacgg\n",
      ">AT2G04890.1 | Symbols: SCL21 | SCARECROW-like 21 | chr2:1720575-1721816 REVERSE LENGTH=1242\n",
      "tacctgttacattctccaag\n",
      "tcacttccatgcacaagatg\n",
      ">AT2G45160.1 | Symbols: LOM1, ATHAM1, SCL27, HAM1 | SCARECROW-LIKE 27, ARABIDOPSIS THALIANA HAIRY MERISTEM 1, HAIRY MERISTEM 1, LOST MERISTEMS 1 | chr2:18618110-18620032 REVERSE LENGTH=1923\n",
      "tacgggaataggaaactttc\n",
      "ctaacatttccaagcagaga\n",
      ">AT3G54220.1 | Symbols: SGR1, SCR | SCARECROW, SHOOT GRAVITROPISM 1 | chr3:20070550-20072625 FORWARD LENGTH=1962\n",
      "taccgccttaggccgctaaa\n",
      "ctaagaacgaggcgtccaag\n",
      ">AT4G17230.1 | Symbols: SCL13 | SCARECROW-like 13 | chr4:9661218-9662807 REVERSE LENGTH=1590\n",
      "tacgtttgcagagtctttgt\n",
      "ctagtttgggtttggcttcc\n",
      ">AT4G36710.1 | Symbols: HAM4, SCL15, AtHAM4 | Scarecrow-like 15, Arabidopsis thaliana HAIRY MERISTEM 4 | chr4:17306060-17307520 FORWARD LENGTH=1461\n",
      "tacttttagggacgtagtag\n",
      "ctaaaaccgccaagctgatg\n",
      ">AT5G41920.1 | Symbols: AtSCL23, SCL23 | SCARECROW-LIKE 23 | chr5:16779982-16781199 FORWARD LENGTH=1218\n",
      "tactggtgttttgcgtgtct\n",
      "tcaatcgaacggctgagatt\n",
      ">AT5G52510.1 | Symbols: SCL8 | SCARECROW-like 8 | chr5:21307196-21309118 FORWARD LENGTH=1923\n",
      "taccttagaccaaagaggcc\n",
      "ttaacgccaagcggatgcga\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "###  Cleaning the fasta file so that is easier to work with it \n",
    "def clean_fasta(input_file, output_file):  # to clean the \"\\n\" out of the DNA sequences but not from the headers fo the sequences or the end of the DNA sequence\n",
    "    with open(input_file, 'r') as f_in, open(output_file, 'w') as f_out:\n",
    "        seq = ''\n",
    "        for line in f_in:\n",
    "            if line.startswith('>'):  # if this line is a header line, write out the previous sequence (if any), and write the header line to the output file\n",
    "                if seq:\n",
    "                    f_out.write(re.sub(r'\\n', '', seq) + '\\n')\n",
    "                f_out.write(line)\n",
    "                seq = ''\n",
    "            else:  \n",
    "                seq += line.strip()  # if this line is a sequence line, add it to the current sequence\n",
    "        f_out.write(re.sub(r'\\n', '', seq) + '\\n')  # write out the last sequence\n",
    "\n",
    "clean_fasta('A_thaliana_cDNA_seqs.txt', \"Clean_cDNA_seqs.txt\")\n",
    "clean_seqs_open = open(\"Clean_cDNA_seqs.txt\", \"r\")  # open the new file to start with the core of the task \n",
    "clean_seqs = clean_seqs_open.readlines()\n",
    "clean_seqs_open.close()\n",
    "print(len(clean_seqs))  # to check how many arguments are in the list and save the names of the sequences in a list\n",
    "seq_names = clean_seqs[::2]  # list with the names of every sequence\n",
    "seq_list = clean_seqs[1::2]  # list with every sequence\n",
    "\n",
    "### Creating the primers from the sequences ###\n",
    "forward_primers_list = []  # list to hold every forward primer\n",
    "reverse_primers_list = []  # list to hold every reverse primer\n",
    "for seq in seq_list:\n",
    "    pure_seq = seq.replace(\"\\n\",\"\")  # erasing the \"n\\\" character to avoid slicing errors \n",
    "    forward_bit = pure_seq[0:20]\n",
    "    forward_a = forward_bit.replace(\"A\",\"t\")\n",
    "    forward_ac = forward_a.replace(\"C\",\"g\")\n",
    "    forward_act = forward_ac.replace(\"T\",\"a\")\n",
    "    forward_primer = forward_act.replace (\"G\",\"c\")\n",
    "    forward_primers_list.append(forward_primer + \"\\n\")  # appending the complement of the 20 first nucleotides to the forward primers list \n",
    "    reverse_bit = pure_seq[-20:]\n",
    "    reversed_bit = reverse_bit[::-1]\n",
    "    reverse_a = reversed_bit.replace(\"A\",\"t\")\n",
    "    reverse_ac = reverse_a.replace(\"C\",\"g\")\n",
    "    reverse_act = reverse_ac.replace(\"T\",\"a\")\n",
    "    reverse_primer = reverse_act.replace(\"G\",\"c\")\n",
    "    reverse_primers_list.append(reverse_primer+ \"\\n\")  # appending the reversed complement of the 20 last nucleotides to the reverse primers list\n",
    "print(\"This is the list for the forward primers\" + str(forward_primers_list))\n",
    "print(\"This is the list for the reverse primers\" + str(reverse_primers_list) + \"\\n\")\n",
    "\n",
    "### Checking the melting temperature ###\n",
    "# Formula to be used: Melting point = 2*(A_counts+T_counts) + 4*(G_counts+C_counts)\n",
    "\n",
    "def forward_melting_point():    \n",
    "    forward_melting_points = []\n",
    "    for primer in forward_primers_list:\n",
    "        numFA = primer.count('a')\n",
    "        numFT = primer.count('t')\n",
    "        numFG = primer.count('g')\n",
    "        numFC = primer.count('c')\n",
    "        melting_point = 2 * (numFA + numFT) + 4 * (numFG + numFC)\n",
    "        forward_melting_points.append(melting_point)\n",
    "    return forward_melting_points\n",
    "\n",
    "forward_melting_points = forward_melting_point()\n",
    "\n",
    "def reverse_melting_point():\n",
    "    reverse_melting_points = []\n",
    "    for primer in reverse_primers_list:\n",
    "        numRA = primer.count('a')\n",
    "        numRT = primer.count('t')\n",
    "        numRG = primer.count('g')\n",
    "        numRC = primer.count('c')\n",
    "        melting_point = 2 * (numRA + numRT) + 4 * (numRG + numRC)\n",
    "        reverse_melting_points.append(melting_point)\n",
    "    return reverse_melting_points\n",
    "\n",
    "reverse_melting_points = reverse_melting_point() \n",
    "\n",
    "### Cheking if melting temperatures are between 55 and 62 degrees and if the difference between primers melting temperatures is less or equal to 4 degrees\n",
    "position = -1\n",
    "forward_positions_above_temp = []\n",
    "forward_positions_below_temp = []\n",
    "for i in forward_melting_points:\n",
    "    position = position + 1\n",
    "    if i > 62:\n",
    "        forward_positions_above_temp.append(position)\n",
    "    elif i < 55:\n",
    "        forward_positions_below_temp.append(position)\n",
    "reverse_positions_above_temp = []\n",
    "reverse_positions_below_temp = []\n",
    "for i in reverse_melting_points:\n",
    "    position = position + 1\n",
    "    if i > 62:\n",
    "        reverse_positions_above_temp.append(position)\n",
    "    elif i < 55:\n",
    "        reverse_positions_below_temp.append(position)\n",
    "if len(forward_positions_above_temp) > 0:\n",
    "    print(\"This positions from the forward primers need to be revised because of high temperature: \" + str(forward_positions_above_temp))\n",
    "elif len(forward_positions_below_temp) > 0:\n",
    "    print(\"This positions from the forward primers need to be revised because of low temperature: \" + str(forward_positions_below_temp) + \"\\n\") # we need to rise this temp manually\n",
    "elif len(reverse_positions_above_temp) > 0:\n",
    "    print(\"This positions from the reverse primers need to be revised because of high temperature: \" + str(reverse_positions_above_temp))\n",
    "elif len(reverse_positions_above_temp) > 0:\n",
    "    print(\"This positions from the reverse primers need to be revised because of low temperature: \" + str(reverse_positions_below_temp))\n",
    "for i in forward_positions_below_temp:  # to rise the melting temperature of the primers\n",
    "    first_index = forward_primers_list[i].find(\"a\")  # find the first occurrence of the character\n",
    "    second_index = forward_primers_list[i].find(\"a\", first_index + 1)\n",
    "    mod_string = forward_primers_list[i][:second_index] + forward_primers_list[i][second_index:].replace(\"a\", \"g\", 2)  # changing 2 \"a\" for \"g\" to rise the melting point\n",
    "    forward_primers_list[i] = mod_string\n",
    "\n",
    "### Calculating again the melting point to check the changes made\n",
    "forward_melting_points = forward_melting_point()\n",
    "reverse_melting_points = reverse_melting_point()\n",
    "print(\"Forward melting points are:\" + str(forward_melting_points))\n",
    "print(\"Reverse melting points are:\" + str(reverse_melting_points) + \"\\n\")\n",
    "\n",
    "### Checking temperature difference between primers\n",
    "flag = False\n",
    "for i in range(12):\n",
    "    if abs(forward_melting_points[i] - reverse_melting_points[i]) > 4:\n",
    "        print(\"Melting point needs to be checked\")\n",
    "        flag = True\n",
    "if flag == False:\n",
    "    print(\"Difference between melting temperatures is optimal\" + \"\\n\")\n",
    "\n",
    "### Export the primers with their corresponding headers into one file\n",
    "primers_list = []\n",
    "for i in range (0,12):\n",
    "    primers_list.append(seq_names[i])\n",
    "    primers_list.append(forward_primers_list[i])\n",
    "    primers_list.append(reverse_primers_list[i])\n",
    "\n",
    "def write_fasta(filename, data):\n",
    "    with open(filename, 'w') as f:\n",
    "        for entry in data:\n",
    "            \n",
    "            header, sequence = entry.split('\\n', 1)  # split the header and sequence\n",
    "            \n",
    "            f.write(header + '\\n')  # write the header to the file\n",
    "            \n",
    "            f.write(sequence.replace('\\n', ''))  # write the sequence to the file, removing any newline characters\n",
    "write_fasta(\"primers_file.txt\", primers_list)\n",
    "with open(\"primers_file.txt\")as final_output:\n",
    "    print(\"The primers for each sequence are:\" + \"\\n\" + str(final_output.read()))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "570feb405e2e27c949193ac68f46852414290d515b0ba6e5d90d076ed2284471"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
